imageio/plugins/rawpy.py

```diff
@@ -0,0 +1,186 @@
+                if request._uri_type == URI_BYTES:
+                    raise InitializationError(
+                        "RawPy can not read the provided bytes."
+                    ) from None
+                else:
+                    raise InitializationError(
+                        f"RawPy can not read {request.raw_uri}."
+                    ) from None
+        elif request.mode.io_mode == IOMode.write:
+            raise InitializationError(
+                "RawPy does not support writing."
+            ) from None
+
+    def close(self) -> None:
+        if self._image_file:
+            self._image_file.close()
+
+        self._request.finish()
+
+    def read(self, *, index: int = 0, **kwargs) -> np.ndarray:
+        """Read Raw Image.
+
+        Returns
+        -------
+        nd_image: ndarray
+            The image data
+        """
+
+        nd_image: np.ndarray
+
+        try:
+            nd_image = self._image_file.postprocess(**kwargs)
+        except Exception:
+            pass
+        
+        nd_image = nd_image[None, ...]
+        return nd_image
+
+    def write(
+        self, 
+        ndimage: Union[ArrayLike, List[ArrayLike]]
+    ) -> Optional[bytes]:
+        """RawPy does not support writing.
+        """
+        raise NotImplementedError()
+
+    def iter(self) -> Iterator[np.ndarray]:
+        """Load the image.
+
+        Returns
+        -------
+        nd_image: ndarray
+            The image data
+        """
+
+        try:
+            yield self.read()
+        except Exception:
+            pass
+
+    def metadata(
+        self, index: int = None, exclude_applied: bool = True
+    ) -> Dict[str, Any]:
+        """Read ndimage metadata.
+
+        Parameters
+        ----------
+        exclude_applied : bool
+            If True, exclude metadata fields that are applied to the image while
+            reading. For example, if the binary data contains a rotation flag,
+            the image is rotated by default and the rotation flag is excluded
+            from the metadata to avoid confusion.
+
+        Returns
+        -------
+        metadata : dict
+            A dictionary of format-specific metadata.
+
+        """
+
+        metadata = {}
+
+        ImageSize = self._image_file.sizes
+
+        # TO DO: Find out what information needs to be included below and refine accordingly.
+        metadata["black_level_per_channel"] = self._image_file.black_level_per_channel
+        metadata["camera_white_level_per_channel"] = self._image_file.camera_white_level_per_channel
+        metadata["color_desc"] = self._image_file.color_desc
+        #metadata["color_matrix"] = self._image_file.color_matrix
+        metadata["daylight_whitebalance"] = self._image_file.daylight_whitebalance
+        metadata["dtype"] = self._image_file.raw_image.dtype
+        metadata["flip"] = ImageSize.flip
+        metadata["num_colors"] = self._image_file.num_colors
```

Comment:
Done. Can you please also tell me why we should remove it if something is consumed by postprocess? Thanks!

Reply:
Because it is less confusing to the user. Take the `flip` metadata as an example. Suppose I do

```
img = iio.imread("foo.dng")
meta = iio.immeta("foo.dng")
```

What does the value of `meta["flip"]` mean here? Do I have to manually flip `img` to restore it to its original form, or should I ignore `flip` because ImageIO flipped it already? If both the user and ImageIO flip the image (or nobody flips) the processing is wrong, but there is no way of knowing without knowing the plugin's internal behavior.

I need to know how to restore the image to its original form including the application of any processing steps that apply after decoding (white balance, flip, rotation, gamma correction, ...). Some plugins apply these by default, others don't. So, to provide a feedback mechanism to the user on what has been applied and what hasn't we hide metadata fields that are applied by default - hence `exclude_applied=True`.

However, some programs need to inspect the _raw_ metadata in the file regardless of the processing applied by default. An example of such a use-case is a programs that coordinates an image processing pipelines. A program might decide that it wants to apply special correction to images that are over-exposed (too bright). In this scenario it needs to know the value of, say, `white_level` to make the correct decision and uses `exclude_applied=False` to get the metadata applicable to the raw file - which we would hide by default because the plugin applies white balance correction by default.

Reply:
Thank you so much for the explanation!